Authorizations via engines & two examples

deivid-rodriguez opened the pull request:
#### :tophat: What? Why?

This PR implements __multi-step__ and __multi-side__ authorizations via engines as suggested in #1871. It keeps support _as is_ for simpler authorizations via rectify forms but introduces the possibility to implement more complex scenarios by declaring a manifest with a engine and (optionally) an admin engine. I've called these new authorization methods "workflow authorizations" as opposed to "handler authorizations".

The manifest is currently very simple and just allows to register the verification workflow and its engines:

https://github.com/decidim/decidim/blob/44dd94490c35340787c968a8126cc1802f837cff/decidim-verifications/lib/decidim/verifications/id_documents/workflow.rb#L1-L6

I'm not sure whether these new authorization engines should live in a separate gem, in a separate repo, namespaced inside `decidim-{core,admin}`? For now I've gone with the first option but I'm open to suggestions.

As the first examples, I've added authorizations via identity document upload as explained in #1815, and authorizations via code by postal mail as explained in #2030.

#### :pushpin: Related Issues
- Fixes #1815.
- Fixes #1871.
- Fixes #2030.
- Sits on top of #2019.

#### :clipboard: Subtasks

##### Common
- [x] Introduce pending verifications. 
- [x] Implement authorization workflows.
- [x] Get "form authorizations" and "workflow authorizations" play nice together.

##### Identity documents
- [x] Identiy document upload.
- [x] Identity document verification.
- [x] Identity document rejection (allows uploading the documents again).

##### Letters
- [x] Address form for postal mail verification.
- [x] Code confirmation for postal mail verification.
- [x] Admin UI to mark letters as sent.

##### Ideas for future iterations (will migrate to separate tickets if accepted as improvements)

- [ ] Extract verification of authorizations to a separate model so it can hold more information about the verification (verification time, user who verified the user if any, ...).  
- [ ] Allow customizing which fields of the identity document get checked via a manifest setting.
- [ ] Allow multiple document upload?
- [ ] Allow batch downloading of letters with codes (in pdf) ready to be sent.
- [ ] Notifications.
- [ ] Unique hashed verification code deterministically generated from user data & application secrets, so letters can be sent out of the platform and users can get verified with it as soon as they get registered https://github.com/decidim/decidim/issues/2030#issuecomment-336477525
- [ ] Reasons for identity document rejection.

### :camera: Screenshots (optional)

#### Enabling authorization methods for an organization
![enableorgauthorizations](https://user-images.githubusercontent.com/2887858/31799778-7e9cf2bc-b53c-11e7-9750-c9aa522b14f6.gif)

#### Enabling authorization methods for specific actions
![enableactionauthorizations](https://user-images.githubusercontent.com/2887858/31799788-88e23a84-b53c-11e7-8f15-246eba64fd56.gif)

#### Verification with identity documents
![authorizingwithdocs](https://user-images.githubusercontent.com/2887858/31799792-91fa10ba-b53c-11e7-8a50-7c2edf0749bb.gif)

#### Verification with postal letter
![authorizingwithletter](https://user-images.githubusercontent.com/2887858/31799839-d88cdb16-b53c-11e7-8eb7-cfe199d0f5e0.gif)

#### :ghost: GIF
![immaterialimpracticalduckbillplatypus-small](https://user-images.githubusercontent.com/2887858/31433509-e26a6884-ae79-11e7-9092-72668a26c99f.gif)

deivid-rodriguez commented:
Hei! This should be ready now for an initial review. I finally added "postal letter verifications" as well in here. Let me know if you want it splitted away, or if it's fine.

I've updated the PR's descriptions and I'll be posting some more screenshots later today.

josepjaume commented:
Hi @deivid-rodriguez - this looks very promising. Before reviewing the code though, there's a tricky scenario you maybe didn't have into account: managed users. Managed users can't be impersonated without re-entering the user's data, which looks kind of incompatible with "deferred" authorizations.

Maybe we could include that information (deferred authorizations) in the manifest in order to exclude them from managed users? It doesn't seem trivial to manage it in a different way.

Thanks!

deivid-rodriguez commented:
> Hi @deivid-rodriguez - this looks very promising. Before reviewing the code though, there's a tricky scenario you maybe didn't have into account: managed users. Managed users can't be impersonated without re-entering the user's authorization data, which looks kind of incompatible with "deferred" (engine?) authorizations.

Yes, I did! :smile: See https://github.com/decidim/decidim/issues/1871#issuecomment-333113815. My solution is to only use direct authorization handlers for impersonations.

deivid-rodriguez commented:
> Maybe we could include that information (deferred authorizations) in the manifest in order to exclude them from managed users? It doesn't seem trivial to manage it in a different way.

Regarding the implementation, at the moment authorization handlers are not registered through a manifest but just like they are registered now through a global configuration. So to figure out the "authorization" type I "auto-detect" this:
* If registered through manifest, it's a workflow (deferred) method.
* If constantizable to a class, then it's a handler (direct) method.

It's this magic here: https://github.com/decidim/decidim/blob/ef12bfba2329e4574f9a66c0bceecf6c6b157df5/decidim-verifications/lib/decidim/verifications/adapter.rb#L1-L56.

Looking backwards I should've probably unify everything under a manifest. I wanted to keep backwards compatibility in this first MVP, but it was probably the wrong choice.

deivid-rodriguez commented:
Some heads up here, I renewed the screenshots. While recording them, I found several bugs and improvements which I've fixed and added. There's some general fixes in here that I could extract out of this PR if you wanted me to.

mrcasals commented:
@deivid-rodriguez screenshots look superawesome! I'll try to find some time today to review the PR! 😄

deivid-rodriguez commented:
@mrcasals Do you see them as blurry as I do? The screen recording process I'm using not only quasi-freezes my laptop but also doesn't generate great gifs. But things should look better in real life :smiley:

mrcasals commented:
Ah yes, GIFs look blurry, but that's a problem with GIFs I guess, so we can't do much about it...

mrcasals commented:
@deivid-rodriguez with this PR, does this mean we can extend decidim with new authorizations bundled in a gem? I'm thinking about the Spanish IDs (DNI, NIE, passport) we're using in Barcelona/Terrassa/Sant Cugat. They are mostly the same (census authorization), they mostly use the same fields, and only change how to check the values to an API. 

Would we be able to extract this in a gem?

josepjaume commented:
Awsome work! Heads up @mrcasals - authorizations can be already bundled in
a gem :)

On Oct 20, 2017 11:58, "Marc Riera" <notifications@github.com> wrote:

> @deivid-rodriguez <https://github.com/deivid-rodriguez> with this PR,
> does this mean we can extend decidim with new authorizations bundled in a
> gem? I'm thinking about the Spanish IDs (DNI, NIE, passport) we're using in
> Barcelona/Terrassa/Sant Cugat. They are mostly the same (census
> authorization), they mostly use the same fields, and only change how to
> check the values to an API.
>
> Would we be able to extract this in a gem?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/decidim/decidim/pull/2024#issuecomment-338162606>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAG0ggcRYZDuA_5VyDvX0wIPUE_MG7lfks5suG7EgaJpZM4P1M0d>
> .
>

mrcasals commented on decidim-verifications/README.md:
```diff
@@ -0,0 +1,96 @@
+  authorization method they require. For example, a decidim instance might
+  choose to require census authorization to create proposals, but a fully
+  verified address via a verification code for voting on proposals.  which
+  authorization.  methods to use to let users get verified, and
```
This last sentence looks weird.

deivid-rodriguez replied:
Oooops, I'll fix this.

mrcasals commented on decidim-verifications/config/locales/en.yml:
```diff
@@ -31,21 +31,31 @@ en:
               verification_number: 'Verification #%{n}'
+          rejections:
+            create:
+              success: Verification rejected. User will be prompted to ammend her documents
```
`s/ammend/amend/`

mrcasals commented:
@deivid-rodriguez über-nice work! I think I'll need to work on adding a new authorization method in order to fully understand how to do that, because it looks a bit complicated, but that's OK.

Should we move the authorizations we have in core to this new engine? They might be simple enough, but this way we keep similar things together.

Apart from that, really nice work, thanks! ❤️

deivid-rodriguez commented:
> @deivid-rodriguez über-nice work! I think I'll need to work on adding a new authorization method in order to fully understand how to do that, because it looks a bit complicated, but that's OK.

Yes, it is a bit complicated... In a future iteration I think we could unify everything under a manifest like @josepjaume intended / suggested. This will be simpler implementation-wise and towards users since we'll have a single way of registering (and documenting) authorization methods.

Something like:

```ruby
 Decidim::Verifications.register_verification_method(:id_documents) do |verification| 
   verification.engine = Decidim::Verifications::IdDocuments::Engine 
   verification.admin_engine = Decidim::Verifications::IdDocuments::AdminEngine 
 end
```
or, for current handlers,

```ruby
 Decidim::Verifications.register_verification_method(:barcelona_census) do |verification| 
   verification.form = Decidim::Barcelona::CensusHandler 
 end
```

However, I think we can release this way for now, since it allows us to start experimenting with this new authorization methods without breaking compatibility with the old ones (and thus not having to deal yet with data migrations, and so on).

deivid-rodriguez commented:
> Should we move the authorizations we have in core to this new engine? They might be simple enough, but this way we keep similar things together.

Absolutely! I scheduled myself to do this, but I forgot! Assuming you mean under the new "gem" (note that there's one "engine" per authorization method, or even two if the authorization method needs admin intervention).

deivid-rodriguez commented:
@mrcasals I'm now fixing the docs in the README and I see that by "complicated" you meant the implementation of the engine itself for a new authorization method. Yes, there's a few rules one needs to follow to play nice with `decidim` in terms of entry points of the engine and how _partial authorizations_ should look. I couldn't find easy ways around this, but we can of course iterate and try to make things as simple as possible in the future.

mrcasals commented:
> @mrcasals I'm now fixing the docs in the README and I see that by "complicated" you meant the implementation of the engine itself for a new authorization method.

@deivid-rodriguez no problem at all if it's complicated, I mean, this shouldn't be blocking.

> Absolutely! I scheduled myself to do this, but I forgot! Assuming you mean under the new "gem" (note that there's one "engine" per authorization method, or even two if the authorization method needs admin intervention).

Yes sorry, I meant we should move the current authorization methods into `decidim-verifications` 😄

deivid-rodriguez commented:
By the current authorization methods, you mean the "ExampleAuthorizationHandler"? It's a bit difficult because that handler is a template for the generator. Maybe I should move it with the rest of the code, I don't see a big benefit in having it as a template.

mrcasals commented:
@deivid-rodriguez you're right, sorry 🤦‍♂️ for some reason I thought we had some real implementations there. We could at most extract the dummy one (the one that needs to end with `X` to be valid), but that's all.

Again, supernice work on this! 😄

deivid-rodriguez commented:
> @deivid-rodriguez you're right, sorry :man_facepalming: for some reason I thought we had some real implementations there. We could at most extract the dummy one (the one that needs to end with X to be valid), but that's all.

Yup, I'll move that with the rest of the code then!

> Again, supernice work on this! :smile:

Thank you so much, your attitude is always so encouraging :heart:

josepjaume commented:
It is supernice work :). Do you mind waiting until monday morning so I can give you my final approval?

deivid-rodriguez commented:
@josepjaume Not all at! :+1:

I've added a few extra changes to further isolate verification code in `decidim-verifications` (see 
804c7d7 and 98a9f66). In particular, I've added a `Decidim::Verifications::Engine` that mounts all register engines, and the regular `authorization` routes. There's more work that can be done in this regard but I feel I should stop now :)

mrcasals commented on decidim-verifications/README.md:
```diff
@@ -59,8 +59,8 @@ Decidim implements two type of authorization methods:
   authorization to succeed, via one or more custom controllers and views. You
-  can create partial `Authorization` records (with the `verified_at` column set
-  to `nil`) and hold partial verification data in the `verification_metadata`
+  can create partial `Authorization` records (with the `verified` column set
```
Did you change it, or were the docs wrong? I'd rather have the `verified_at` field than a boolean 😕

deivid-rodriguez replied:
@mrcasals You're absolutely right! I'm reverting back to the `verified_at` column.

This story is that I started with a datetime column, and then changed to a boolean one because I was running into uniqueness validation problems. However, now that I re-think about it, I don't know why I was insisting on validating the triplet `[:user, :name, :granted(_at)]`, I don't think a user should ever have two authorizations of the same type, so the correct validation is the old one (uniqueness just on `[:user, :name]`)...

deivid-rodriguez replied:
I've added a few more changes including reverting this change and restoring the `verified_at` column.

deivid-rodriguez commented:
Hei! I've rebased this PR and added some more changes:

* I decided to make the API uniform as previously suggested. Now all authorizations are registered similarly. This PR is now backwards incompatible, but I've added data migrations to ease the process.

* Added change log entries.

* Fixed some errors uncovered by recently added specs (#2119).

* Moved some more authorization specific code to `decidim-verifications`.

mrcasals commented:
@deivid-rodriguez CodeClimate seems to be complaining, can you address the issues? :smile:

deivid-rodriguez commented:
Done!

Also, there's several unrelated changes I could extract to separate PR's and some commits I could squash as well. Let me know if that would help with reviewing this!

deivid-rodriguez commented:
I rebased this PR, extracted a few more changes to separate PR's that already got merged, and squashed a few commits. Hopefully build is still passing.

I can extract more things if needed. Specifically, I can divide this into 3 PRs if you wanted me to:

* Generalize verification method registration to the new proposed API.
* Add verification by identity documents.
* Add verification by postal letter code.

Let me know!

deivid-rodriguez commented:
@decidim/developers I rebased this PR, I don't mind to split it now as I mentioned if you prefer!

Also, there's a couple of codeclimate issues, but I don't think it's worth fixing them. However, I don't think I have permissions to ignore them (https://docs.codeclimate.com/docs/disabling-individual-checks).

mrcasals commented:
@deivid-rodriguez now that #2239 is merged you can fix the Codeclimate issues, right? Also, note that there are some conflicts :)

deivid-rodriguez commented:
@mrcasals Yeah, I partially fixed them but in #2239 I also relaxed the check so codeclimate is green now.

However, I bumped into another problem. In this case, I think #2241 should fix it!

deivid-rodriguez commented:
After a couple of detours, this is passing again!

mrcasals commented on decidim-verifications/lib/decidim/verifications/workflows.rb:
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+module Decidim
+  module Verifications
```
Docs?

mrcasals commented on decidim-verifications/lib/decidim/verifications/workflow_manifest.rb:
```diff
@@ -0,0 +1,37 @@
+
+module Decidim
+  module Verifications
+    class WorkflowManifest
```
Docs?

mrcasals commented on decidim-verifications/lib/decidim/verifications/registry.rb:
```diff
@@ -0,0 +1,39 @@
+module Decidim
+  module Verifications
+    #
+    # Takes care of holding and accessing verification methods.
```
A little more docs?

mrcasals commented on decidim-verifications/lib/decidim/verifications/adapter.rb:
```diff
@@ -0,0 +1,74 @@
+    class UnregisteredVerificationManifest < StandardError
+    end
+
+    class Adapter
```
Docs?

mrcasals commented on decidim-verifications/app/presenters/decidim/verifications/postal_letter/authorization_presenter.rb:
```diff
@@ -0,0 +1,41 @@
+module Decidim
+  module Verifications
+    module PostalLetter
+      class AuthorizationPresenter < SimpleDelegator
```
Docs?

deivid-rodriguez commented:
Sure, I'll add docs for those classes on a new commit!

deivid-rodriguez commented:
@mrcasals I added docs!
